api/v3/Order.php

```diff
@@ -127,6 +127,40 @@ function civicrm_api3_order_create(&$params) {
 }
 
 /**
+ * Cancel a Order.
+ *
+ * @param array $params
+ *   Input parameters.
+ *
+ * @return array
+ */
+function civicrm_api3_order_cancel($params) {
+  $contributionStatuses = CRM_Contribute_PseudoConstant::contributionStatus(NULL, 'name');
+  $params['contribution_status_id'] = array_search('Cancelled', $contributionStatuses);
```

Comment:
$params['contribution_status_id'] = array_search('Cancelled', $contributionStatuses); is needed in CRM_Contribute_BAO_Contribution::transitionComponents($params);  

transitionComponents() function checks for status like Cancelled, Failed etc and then updates the status of related components accordingly.